### PR TITLE
buildGoModule: don't inherit build hooks when building go-modules

### PR DESCRIPTION
**Pull request by doronbehar**
Inheriting these hooks in the `.go-modules` derivation can be confusing:
One doesn't expect them to run when generating the fixed output modules
derivation, but only on the main derivation. A `postBuild` hook that
adds some files to $out will cause a very hard to debug issue[1].

This commit adds support for dedicated hooks that will be used only by
the derivation building `.go-modules`.

[1]:
https://discourse.nixos.org/t/cant-update-a-go-package-getting-go-inconsistent-vendoring/27063/6###### Description of changes

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by doronbehar** [comment]
Marked for review because I'm not totally sure if this will affect other packages' go-modules derivations. I presume almost no packages will need further patching. In any case it may be better to review the changes before sprinting over go packages with hooks.

**Review by zowoq — CHANGES_REQUESTED** [review]
This is basically the opposite of https://github.com/NixOS/nixpkgs/pull/212800.

**Comment by doronbehar** [comment]
> This is basically the opposite of #212800.

My change at least proposes to document the behavior somehow. I don't strongly care what it will be, I just don't want to spend hours debugging `buildGoModules` trying to figure out what's wrong with my `vendor/` directory. Note also how #212800 broke some packages.

I think the adaptations these maintainers did were justified never the less. I see this PR as an improvement above #212800.

**Comment by zowoq** [comment]
> > This is basically the opposite of #212800.
> 
> My change at least proposes to document the behavior somehow. I don't strongly care what it will be, I just don't want to spend hours debugging `buildGoModule` trying to figure out what's wrong with my `vendor/` directory. Note also how #212800 broke some packages.
> 
> I think the adaptations these maintainers did were justified never the less. I see this PR as an improvement on top of #212800.

Just to make sure that I'm not misunderstanding, you'd be okay with the current behaviour if it's documented?

So we could drop bb5864d50ec93e70bd3bad3249d1599541e093f8 and just merge 2046e7a0a11e65462c1c3361e92e49699a403d58 by itself?

**Comment by doronbehar** [comment]
> Just to make sure that I'm not misunderstanding, you'd be okay with the current behaviour if it's documented?

I'm open for negotiation :)

> I'd be open to reconsidering the postBuild change from https://github.com/NixOS/nixpkgs/pull/212800 as it's only been a couple of months but I'd be very reluctant to change the preBuild behaviour as it's been like that for almost three years.

My change doesn't remove the _ability_ to change the behavior of the go-modules derivation. It only changes the API. I think it's worth it as the control over the `vendor/` generation should be separated from the control of the main build, IMO.

We can iterate all packages that use `preBuild` (83 all together) and test whether their `preBuild` should be changed to `modPreBuild` - to preserve the `vendorSha256`. I suspect almost all of them won't need such change, have a look at:

```
git grep -A4 "preBuild = ''" $(git grep -l buildGoModule)
```

**Comment by zowoq** [comment]
> My change doesn't remove the ability to change the behavior of the go-modules derivation. It only changes the API.

Yes, I don't want the API to change.

> I'm open for negotiation

I think the compromise would be changing postBuild to modPostBuild, keeping preBuild as it is and documenting it.

**Comment by doronbehar** [comment]
> I think the compromise would be changing postBuild to modPostBuild, keeping preBuild as it is and documenting it.

I understand, but I really think it's feasable to iterate all packages that use `preBuild` and test whether this change actually breaks them.

**Comment by zowoq** [comment]
> I understand, but I really think it's feasable to iterate all packages that use `preBuild` and test whether this change actually breaks them.

I'm not just thinking about nixpkgs, I'm thinking about external usage. I do not see the justification in changing it after it's been like that for three years.

**Comment by doronbehar** [comment]
> I'm not just thinking about nixpkgs, I'm thinking about external usage. I do not see the justification in changing it after it's been like that for three years.

I understand. In the current form of the PR, I also added a comment near where `preBuild = args.preBuild` to perhaps warn against future changes of that section.

**Comment by zowoq** [comment]
LGTM, thanks. Let's give Sandro a bit of time to comment on the change before it gets merged.

File: doc/languages-frameworks/go.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/go.section.md:118]:
```suggestion
Many attributes [controlling the build phase](#variables-controlling-the-build-phase) are respected by both `buildGoModule` and `buildGoPackage`. Note that `buildGoModule` reads the following attributes also when building the `vendor/` go-modules fixed output derivation as well:
```

File: doc/languages-frameworks/go.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/go.section.md:23]:
```suggestion
- `modPostBuild`: Shell commands to run after the build of the go-modules executes `go mod vendor`, and before calculating fixed output derivation's `vendorHash` (or `vendorSha256`). Note that if you change this attribute, you need to update `vendorHash` (or `vendorSha256`) attribute.
```

File: pkgs/build-support/go/module.nix
Comment (SuperSandro2000) [inline pkgs/build-support/go/module.nix:94]:
```suggestion
    preBuild = args.preBuild or args.modPostBuild or "";
```
Maybe we should add the parity attribute for preBuild to be a bit consistent

File: doc/languages-frameworks/go.section.md
Comment (SuperSandro2000) [inline doc/languages-frameworks/go.section.md:125]:
```suggestion
- [`sourceRoot`](#var-stdenv-sourceRoot)
- [`prePatch`](#var-stdenv-prePatch)
- [`patches`](#var-stdenv-patches)
- [`patchFlags`](#var-stdenv-patchFlags)
- [`postPatch`](#var-stdenv-postPatch)
- [`preBuild`](#var-stdenv-preBuild)
```
I would rather sort this by the order the phases are executed than alphabetical

File: pkgs/build-support/go/module.nix
Comment (doronbehar) [reply pkgs/build-support/go/module.nix:94]: > Maybe we should add the parity attribute for preBuild to be a bit consistent

I'm not sure I understand. Perhaps you meant to also respect `args.modPreBuild`? and not `args.modPostBuild`? When setting `preBuild`..

Comment (doronbehar) [reply pkgs/build-support/go/module.nix:94]: Friendly ping @SuperSandro2000 .

Comment (SuperSandro2000) [reply pkgs/build-support/go/module.nix:94]: Yeah, exactly. Forgot to change that when copying

**Pull request merged**
